lib/exmeralda/regenerations.ex

```diff
@@ -0,0 +1,193 @@
+defmodule Exmeralda.Regenerations do
+  @moduledoc """
+  Context around regenerating messages
+  """
+  require Logger
+
+  import Ecto.Query
+  alias Exmeralda.Repo
+  alias Exmeralda.Chats
+  alias Exmeralda.Regenerations.Server, as: RegenerationServer
+  alias Exmeralda.Regenerations.JSON, as: JSONFormatter
+  alias Ecto.Multi
+  alias Exmeralda.Chats.{Message, Session, GenerationEnvironment}
+
+  @type regenerate_opts :: [download: boolean(), download_path: String.t()]
+  @spec regenerate_messages([Message.id()], GenerationEnvironment.id()) :: map()
+  @spec regenerate_messages([Message.id()], GenerationEnvironment.id(), regenerate_opts()) ::
+          map()
+  def regenerate_messages(message_ids, generation_environment_id, opts \\ []) do
+    if Mix.env() == :prod do
+      raise "sorry this only works locally for now"
+    end
+
+    download? = Keyword.get(opts, :download, false)
+
+    Logger.info("🏁 Initiating the regeneration. This can take some time...")
+    Process.flag(:trap_exit, true)
+
+    {:ok, pid} =
+      GenServer.start(RegenerationServer, %{
+        message_ids: message_ids,
+        generation_environment_id: generation_environment_id
+      })
+
+    ref = Process.monitor(pid)
+    state = GenServer.call(pid, :regenerate)
+
+    if state.regenerated_messages == %{} do
+      GenServer.stop(pid, {:shutdown, :nothing_to_regenerate})
+      Logger.info("💥 Nothing to regenerate! Reason:")
+    end
+
+    receive do
+      {:DOWN, ^ref, :process, ^pid, {:shutdown, :nothing_to_regenerate}} ->
+        Logger.info("Skipped Messages:")
+        state.skipped_messages
+
+      {:DOWN, ^ref, :process, ^pid, :normal} ->
+        Logger.info("✅ Regeneration finished")
+
+        result = %{
+          regenerated_messages:
+            Enum.map(state.regenerated_messages, fn {k, v} ->
+              {k, Map.take(v, [:assistant_message_id])}
+            end),
+          skipped_messages: state.skipped_messages
+        }
+
+        if download? do
+          Logger.info("🎁 Results: #{inspect(result)}")
+
+          download(
+            Enum.map(state.regenerated_messages, fn {_k, v} -> v.assistant_message_id end),
+            opts
+          )
+        else
+          Logger.info("🎁 Results:")
+          result
```

Comment:
this pops up in the wrong order:

<img width="678" height="171" alt="Screenshot 2025-09-10 at 12 33 01" src="https://github.com/user-attachments/assets/28ece1ff-da6b-4468-a042-8c38a594f1c8" />

Maybe we can just remove line 67?